### ISSUE TITLE
Make it possible to set the number of ticks per beat

### ISSUE THREAD

**Issue by flesler** [dingram/jsmidgen#11]
from https://github.com/dingram/jsmidgen/issues/10#issuecomment-158172118

**Comment by jarmitage** [comment]
+1 please!

**Comment by walmik** [comment]
Hello @dingram I intend to add this feature but need some guidance. At a high level I was thinking of adding a method on File.prototype that takes a number (eg 1000) and sets the File.HDR_SPEED to `"\x03\xe8"`

I played with a few ways to write this method and am not sure if any of it is the right approach. My issue is, I m ending up doing some explicit String manipulation to create the correct format. I wish there was a better way to do this (which is why I m wondering if you could help me identify it)

Starting with the extremely explicit:
```
function a(num) {
	var hex;
	var hdrSpeed;
       /**
	 * Early return in case of invalid input
	 * Invalid input scenarios:
	 * - missing `num` argument
	 * - num is not a number
	 * - num is a number but is out of bounds (HDR_SPEED is 2 bytes only)
	 */
	if (!num || typeof num !== 'number' || Buffer.byteLength(num, 'hex') > 2) {
		return;
	}

	hex = num.toString(16);
	switch (hex.length) {
		case 1:
			hdrSpeed = '\\x00\\x0' + hex;
			break;

		case 2:
			hdrSpeed = '\\x00\\x' + hex;
			break;

		case 3:
			hdrSpeed = '\\x0' + hex[0] + '\\x' + hex[1] + hex[2];
			break;

		case 4:
			hdrSpeed = '\\x' + hex[0] + hex[1] + '\\x' + hex[2] + hex[3];
			break;
	}

	return hdrSpeed; // or directly update File.HDR_SPEED here
}
```

The next one is not all that explicit, but I still need to manipulate a String explicitly (which I m not happy with)
```
function b(integer) {
	var length = Math.ceil((Math.log(integer)/Math.log(2))/8); // How much byte to store integer in the buffer
	var buffer = new Buffer(length);
	var arr = []; // Use to create the binary representation of the integer
	var retVal;

	while (integer > 0) {
		var temp = integer % 2;
		arr.push(temp);
		integer = Math.floor(integer/2);
	}

	var counter = 0;
	var total = 0;

	for (var i = 0; i < arr.length; i++) {
		if (counter % 8 == 0 && counter > 0) { // Do we have a byte full ?
			buffer[length - 1] = total;
			total = 0;
			counter = 0;
			length--;
		}

		// User arr element by converting it to number first
		if (+arr[i] == 1) { // bit is set
			total += Math.pow(2, counter);
		}
		counter++;
	}

	buffer[0] = total;

	return buffer.toString('hex');
	// And then use the string thus obtained and manipulate it just like in the previous example
}
```

And then a version where I m just simplifying the previous one a little (but still needing to do a String manipulation)

```
function c(ticks) {
	var validByteLength = 2;		// HDR_SPEED is 2 bytes only
	var hexBuffer, requiredByteLength, bitArr, counter = 0, total = 0, byteLength;

	/**
	 * Early return in case of invalid input
	 * Invalid input scenarios:
	 * - missing `ticks` argument
	 * - ticks is not a number
	 * - ticks is a number but is out of bounds
	 */
	if (!ticks || typeof ticks !== 'number') {
		return;
	}

	// Get number of bytes required to store the ticks
	requiredByteLength = Buffer.byteLength(ticks, 'hex');

	// Consider it out of bounds in case its more than the valid byte length
	if (requiredByteLength > validByteLength) {
		return;
	}

	hexBuffer = new Buffer(requiredByteLength);
	bitArr = ticks.toString(2).split('').reverse();	// This will create an array of stringed 1s and 0s
	byteLength = requiredByteLength;

	bitArr.forEach(function(bit) {
		// Check if we have a full byte
		if (counter % 8 === 0 && counter > 0) {
			// console.log('here counter modulo 8 is 0');
			hexBuffer[byteLength - 1] = total;
			counter = 0;
			total = 0;
			byteLength--;
		}

		// bit is a {String} 0 or 1
		if (+bit === 1) {
			total += Math.pow(2, counter);
		}

		counter++;
	});
	
	hexBuffer[0] = total;

        return hexBuffer.toString('hex');
        // And then I need to manipulate this string to create the correct format
}
```

So all of these approaches make me do String manipulation which makes me favor option `a` as its probably faster. Having said that, I m not so sure bout this explicit approach. Please let me know your comments/thoughts. And yes, Merry Christmas!

:)

**Comment by walmik** [comment]
Created a PR for this feature https://github.com/dingram/jsmidgen/pull/12

**Comment by dingram** [comment]
Merged and published in v0.1.4.
